osu.Game/Screens/Edit/Compose/Components/CircularDistanceSnapGrid.cs

```diff
@@ -100,5 +102,40 @@ public override (Vector2 position, double time) GetSnappedPosition(Vector2 posit
 
             return (snappedPosition, snappedTime);
         }
+
+        private class Ring : CircularProgress
+        {
+            [Resolved]
+            private IDistanceSnapProvider snapProvider { get; set; }
+
+            [Resolved(canBeNull: true)]
+            private EditorClock editorClock { get; set; }
+
+            private readonly HitObject referenceObject;
+
+            private readonly Color4 baseColour;
+
+            public Ring(HitObject referenceObject, Color4 baseColour)
+            {
+                this.referenceObject = referenceObject;
+
+                Colour = this.baseColour = baseColour;
+
+                Current.Value = 1;
+            }
+
+            protected override void Update()
+            {
+                base.Update();
+
+                if (editorClock == null)
+                    return;
+
+                float distanceForCurrentTime = snapProvider.DurationToDistance(referenceObject, editorClock.CurrentTime - referenceObject.GetEndTime());
```

Comment:
Feel we may want to not pass the `referenceObject` in but instead just `StartTime`...

Reply:
Actually we'll still have to pass `referenceObject` either way, that's why I used `GetEndTime` instead. Either or works I think.

Reply:
ah, let's leave it then